src/trace/implementations/mod.rs

```diff
@@ -167,3 +168,251 @@ impl<T: Columnation> RetainFrom<T> for TimelyStack<T> {
+        type Item = T;
+        fn push(&mut self, item: <Self::Item as ToOwned>::Owned) where Self::Item: ToOwned {
+            self.copy(item.borrow());
+        }
+        fn copy(&mut self, item: &T) {
+            self.copy(item);
+        }
+        fn copy_slice(&mut self, slice: &[<Self::Item as ToOwned>::Owned]) where Self::Item: ToOwned {
+            self.reserve_items(slice.iter());
+            for item in slice.iter() {
+                self.copy(item);
+            }
+        }
+        fn copy_range(&mut self, other: &Self, start: usize, end: usize) {
+            let slice = &other[start .. end];
+            self.reserve_items(slice.iter());
+            for item in slice.iter() {
+                self.copy(item);
+            }
+        }
+        fn with_capacity(size: usize) -> Self {
+            Self::with_capacity(size)
+        }
+        fn reserve(&mut self, _additional: usize) {
+        }
+        fn merge_capacity(cont1: &Self, cont2: &Self) -> Self {
+            let mut new = Self::default();
+            new.reserve_regions(std::iter::once(cont1).chain(std::iter::once(cont2)));
+            new
+        }
+        fn index(&self, index: usize) -> &Self::Item {
+            &self[index]
+        }
+        fn len(&self) -> usize {
+            self[..].len()
+        }
+    }
+
+    /// A container that accepts slices `[B::Item]`.
+    pub struct SliceContainer<B> {
+        /// Offsets that bound each contained slice.
+        ///
+        /// The length will be one greater than the number of contained slices,
+        /// starting with zero and ending with `self.inner.len()`.
+        pub offsets: Vec<usize>,
+        /// An inner container for sequences of `B` that dereferences to a slice.
+        pub inner: Vec<B>,
+    }
+
+    impl<B: Default> BatchContainer for SliceContainer<B>
+    where
+        B: Clone + Sized,
+        [B]: ToOwned<Owned = Vec<B>>,
+    {
+        type Item = [B];
+        fn push(&mut self, item: Vec<B>) where Self::Item: ToOwned {
+            for x in item.into_iter() {
+                self.inner.push(x);
+            }
+            self.offsets.push(self.inner.len());
+        }
+        fn copy(&mut self, item: &Self::Item) {
+            for x in item.iter() {
+                self.inner.copy(x);
+            }
+            self.offsets.push(self.inner.len());
+        }
+        fn copy_slice(&mut self, slice: &[Vec<B>]) where Self::Item: ToOwned {
+            for item in slice {
+                self.copy(item);
+            }
+        }
+        fn copy_range(&mut self, other: &Self, start: usize, end: usize) {
+            for index in start .. end {
+                self.copy(other.index(index));
+            }
+        }
+        fn with_capacity(size: usize) -> Self {
+            Self {
+                offsets: Vec::with_capacity(size),
+                inner: Vec::with_capacity(size),
+            }
+        }
+        fn reserve(&mut self, _additional: usize) {
```

Comment:
Note to @antiguru: we could have a slightly better implementation here to pre-size the local part of self, but I can't recall the available apis out of my head.